### PR TITLE
fix(union): remap sys ids value to make them unique

### PR DESCRIPTION
**Pull request by shcheklein**
Fixes https://github.com/iterative/datachain/issues/1356

## Summary by Sourcery

Remap sys__id in union operations to guarantee uniqueness and avoid overflow, and add a test to prevent duplication in large union persists

Bug Fixes:
- Ensure sys__id values are unique across union branches by remapping them to dense row numbers with branch-specific offsets

Enhancements:
- Introduce internal _remap_sys_id_dense function to compute new sys__id values as row_number()*2 + branch_offset

Tests:
- Add functional test to verify no duplicated rows when persisting large unions across batch boundaries

### CODE REVIEW

File: src/datachain/query/dataset.py [hunk 1]
```diff
@@ -1220,6 +1223,28 @@ def _order_columns(
...
+    The resulting sys__id is computed as:
+      row_number(order by sys__id) * 2 + branch_offset
+
+    This guarantees uniqueness across branches (even vs odd) and avoids
+    arithmetic overflow from scaling arbitrary pre-existing ids.
+    """
+    remapped: list[ColumnElement] = []
+    for col in columns:
+        if getattr(col, "name", None) == "sys__id":
+            rn = sqlalchemy.func.row_number().over(order_by=col)
```
Comment (shcheklein) [inline src/datachain/query/dataset.py]: not sure about this one cc @dreadatour  - do we need reproducible order? or is it fine to drop it completely ... means each time we run we can a differnet sequence after union ... which is probably fine 🤔 ?

File: src/datachain/query/dataset.py [hunk 2]
```diff
@@ -982,6 +982,9 @@ def apply(
 
         columns1, columns2 = _order_columns(q1.columns, q2.columns)
 
+        columns1 = _remap_sys_id_dense(columns1, branch_offset=0)
+        columns2 = _remap_sys_id_dense(columns2, branch_offset=1)
```
Comment (dreadatour) [inline src/datachain/query/dataset.py]: We should also remap `sys__random` field. After `union` there might (will?) be rows with similar `sys__random` field, and we want (?) `sys__random` to be different across all table.

File: src/datachain/query/dataset.py [hunk 3]
```diff
@@ -1220,6 +1223,28 @@ def _order_columns(
     return [[d[n] for n in column_order] for d in column_dicts]
 
 
+def _remap_sys_id_dense(
+    columns: Iterable[ColumnElement], branch_offset: int
+) -> list[ColumnElement]:
```
Comment (dreadatour) [inline src/datachain/query/dataset.py]: I believe this should be part of `warehouse` class implementation to be able to use different default value for different warehouse implementations. E.g. `row_number` for SQLite and `random` for ClickHouse.

File: src/datachain/query/dataset.py (hunk 1 above)
Comment (dreadatour) [reply src/datachain/query/dataset.py]: I believe it is fine to ignore the order.

In ClickHouse order is not guaranteed and it was OK until now. But same time may be (?) someone rely on order as "undocumented feature" and this change might break this behavior.

File: src/datachain/query/dataset.py [hunk 4]
```diff
@@ -1220,6 +1223,28 @@ def _order_columns(
...
+      row_number(order by sys__id) * 2 + branch_offset
+
+    This guarantees uniqueness across branches (even vs odd) and avoids
+    arithmetic overflow from scaling arbitrary pre-existing ids.
+    """
+    remapped: list[ColumnElement] = []
+    for col in columns:
+        if getattr(col, "name", None) == "sys__id":
+            rn = sqlalchemy.func.row_number().over(order_by=col)
+            new_id = (rn * 2) + sqlalchemy.literal(branch_offset)
```
Comment (dreadatour) [inline src/datachain/query/dataset.py]: Just wonder, should be have `2` as function argument with default value?

Something like:
```python
def _remap_sys_id_dense(
    columns: Iterable[ColumnElement],
    branch_offset: int,
    branches_total: int = 2,
) -> list[ColumnElement]:
```

to be able to use more than 2 branches in future.

Otherwise we might want to use `bool` as `branch_offset` (like `even`/`odd` branch), since only two branches are supported here.

**Review by dreadatour — APPROVED** [review]
Thank you for the fix! Looks good to me, approving to unblock, but please, take a look at comments below 🙏

File: src/datachain/query/dataset.py (hunk 3 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: unless it is really needed, why would we want to have to manage two different implementations? is it known that one of them is substantially faster?

File: src/datachain/query/dataset.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: 🤔 do we guarantee sys__random to exist? where do we use it?

Comment (dreadatour) [reply src/datachain/query/dataset.py]: Sorry, it is `sys__rand`, not `sys__random`.

> do we guarantee sys__random to exist?

No 100% guarantee (since this field can be filtered out), but we add both `id` and `rand` sys signals in one place: https://github.com/iterative/datachain/blob/72784953938c23409d0a72d84a6483be0e948929/src/datachain/data_storage/schema.py#L276-L282

> where do we use it?

1. In `train_test_split` here: https://github.com/iterative/datachain/blob/72784953938c23409d0a72d84a6483be0e948929/src/datachain/toolkit/split.py#L72 — we are using pre-baked `random` value from `sys__rand` to get deterministic results.
2. In `apply_steps` here: https://github.com/iterative/datachain/blob/72784953938c23409d0a72d84a6483be0e948929/src/datachain/query/dataset.py#L1397 (but I can not remember any place where we do use `DATACHAIN_QUERY_CHUNK_INDEX` and `DATACHAIN_QUERY_CHUNK_TOTAL`).

File: src/datachain/query/dataset.py (hunk 4 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: it's a simple change - can do it later when / if it is needed

File: src/datachain/query/dataset.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: > but I can not remember any place where we do use DATACHAIN_QUERY_CHUNK_INDEX and DATACHAIN_QUERY_CHUNK_TOTAL)

omg, what is this?

> — we are using pre-baked random value from sys__rand to get deterministic results.

if we use it only there, we should probably just add that column there? it seems heavyhanded to maintain it ... how wound window/ groupby functions like rand() and Clickhouse?

File: src/datachain/query/dataset.py (hunk 3 above)
Comment (dreadatour) [reply src/datachain/query/dataset.py]: This is because ClickHouse is distributed and we can not use autoincremented integers as primary keys.

I am not sure how `row_number()` will works on a large scale in ClickHouse — it should works, I can not see any reasons why not, but same time this may leads to some weird flaky bugs, and it will be really hard to investigate those, especially on a large scale.

Deeper research is needed here IMO, or we can stay on a safe side and use random ints, same as we use it in all other places for `sys__id`.

File: src/datachain/query/dataset.py (hunk 2 above)
Comment (dreadatour) [reply src/datachain/query/dataset.py]: > omg, what is this?

Some leftovers, I guess 🫣

> if we use it only there, we should probably just add that column there?

Then results of `train_test_split` will be not deterministic, if we will add `sys__rand` field there before each run. We can simply use `rand()` SQL function instead in this case. We need to have this `sys__rand` field in each dataset pre-baked, to be able to apply `train_test_split`.

More details in original issue here: https://github.com/iterative/datachain/issues/603

The idea (from very beginning, I guess, I am not sure, we do have this field from the very beginning) was to always have this `sys__rand` field in the DB, just in case we need any random filtering/order/etc

> it seems heavyhanded to maintain it...

It is, yes.

File: src/datachain/query/dataset.py (hunk 3 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: I just don't really want to have custom implementation across project unless it is really really needed

Comment (shcheklein) [reply src/datachain/query/dataset.py]: > I am not sure how row_number() will works on a large scale in ClickHouse 

do you have any reason / suspicion there?

File: src/datachain/query/dataset.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: > The idea (from very beginning, I guess, I am not sure, we do have this field from the very beginning) was to always have this sys__rand field in the DB, just in case we need any random filtering/order/etc

omg, this way way too heavy (and probably doesn't work anyways in many cases) :(

Comment (shcheklein) [reply src/datachain/query/dataset.py]: okay, I guess I will add it, but it is bad design (too heavy if we just use it for split)

Comment (dreadatour) [reply src/datachain/query/dataset.py]: We only use it internally for split, since we already have it, but this field was there from very beginning, also for users to be able to use it if needed. I guess.

File: src/datachain/query/dataset.py (hunk 3 above)
Comment (dreadatour) [reply src/datachain/query/dataset.py]: Me too. Just trying to highlight all possible edge cases.

Let me take a really quick research on row_number, just to be sure?

Comment (shcheklein) [reply src/datachain/query/dataset.py]: yep, thanks. Much appreciated

File: src/datachain/query/dataset.py (hunk 1 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: discussed with @dreadatour . No apparent reason to believe CH breaks on row_number(), we are waiting a response for a bit from CH team, but then we can merge this.

File: src/datachain/query/dataset.py (hunk 3 above)
Comment (dreadatour) [reply src/datachain/query/dataset.py]: It looks like it should be ok to use `row_number` in ClickHouse 👍

Comment (shcheklein) [reply src/datachain/query/dataset.py]: for the record, migrate to the new helper that was used for outer merge as well. It covered sys__rand and it is using backend-specific approaches.

File: src/datachain/query/dataset.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/query/dataset.py]: done with the new helper

File: src/datachain/query/dataset.py [hunk 5]
```diff
@@ -982,18 +982,26 @@ def apply(
 
         columns1, columns2 = _order_columns(q1.columns, q2.columns)
 
+        union_select = sqlalchemy.select(*columns1).union_all(
+            sqlalchemy.select(*columns2)
+        )
+        union_cte = union_select.cte()
```
Comment (shcheklein) [inline src/datachain/query/dataset.py:988]: Note: `cte()` added to break a very long queries since union is nesting more and more and some tests fail on older SQLite with stack overflow

File: .github/workflows/tests-studio.yml [hunk 6]
```diff
@@ -32,7 +32,7 @@ jobs:
           POSTGRES_DB: database
           POSTGRES_HOST_AUTH_METHOD: trust
       clickhouse:
-        image: clickhouse/clickhouse-server:24.8
+        image: clickhouse/clickhouse-server:25.8
```
Comment (shcheklein) [inline .github/workflows/tests-studio.yml:35]: Note: to support `cte()` below

File: tests/unit/lib/test_datachain.py [hunk 7]
```diff
@@ -3686,7 +3686,10 @@ def to_parent(signal: Signal) -> Parent:
         "cnt": "int",
         "sum": "float",
     }
-    assert chain.to_records() == [
+    assert sorted(
```
Comment (shcheklein) [inline tests/unit/lib/test_datachain.py:3689]: Note: required on newer CH version
